src/Controller/ComponentRegistry.php

```diff
@@ -35,35 +34,29 @@ class ComponentRegistry extends ObjectRegistry implements EventDispatcherInterfa
     use EventDispatcherTrait;
 
     /**
-     * The controller that this collection was initialized with.
+     * The controller that this collection is associated with.
      *
-     * @var \Cake\Controller\Controller|null
+     * @var \Cake\Controller\Controller
      */
-    protected ?Controller $_Controller = null;
+    protected Controller $_Controller;
 
     /**
      * Constructor.
      *
-     * @param \Cake\Controller\Controller|null $controller Controller instance.
+     * @param \Cake\Controller\Controller $controller Controller instance.
      */
-    public function __construct(?Controller $controller = null)
+    public function __construct(Controller $controller)
     {
-        if ($controller) {
-            $this->setController($controller);
-        }
+        $this->setController($controller);
```

Comment:
I think we would remove setController() as well since the constructor argument is required.

Reply:
I did consider that but one can set a custom registry instance for the controller using `Controller::components()` and this method is called there which ensures the registry always uses the controller instance it's "attached" to.

Reply:
All of this feels hacky.

Reply:
We could get rid of the ability to set a registry instance using `Controller::components()`, I see no practical use case for it.

Reply:
Removing the setter mode for Controller::components() sounds pretty good to me. I don't think I have ever seen it used. From within a controller user land code could make their own custom registry in a constructor if customization is required.

Reply:
Done